packages/flutter/lib/src/services/text_input.dart

```diff
@@ -671,6 +686,370 @@ class RawFloatingCursorPoint {
+  /// The range that is being replaced can either grow or shrink based on the
+  /// given replacement text.
+  /// A replacement can occur in cases such as auto-correct, suggestions, and
+  /// when a selection is replaced by a single character.
+  /// {@endtemplate}
+  replacement,
+
+  /// {@template flutter.services.TextEditingDeltaEquality}
+  /// The delta is not modifying the text. There are potentially selection and
+  /// composing region updates in the delta that still need to be applied to your
+  /// text model.
+  /// A situation where this delta would be created is when dragging the selection
+  /// handles. There are no changes to the text, but there are updates to the selection
+  /// and potentially the composing region as well.
+  /// {@endtemplate}
+  equality,
+}
+
+/// A mixin for manipulating a string of text.
+mixin TextEditingDeltaUtils {
+  /// Replaces a range of text in the original string with the text given in the
+  /// replacement string.
+  String replace(String originalText, String replacementText, int start, int end) {
+    final String textStart = originalText.substring(0, start);
+    final String textEnd = originalText.substring(end, originalText.length);
+    final String newText = textStart + replacementText + textEnd;
+    return newText;
+  }
+}
+
+/// A structure representing a granular change that has occured to the editing
+/// state as a result of text editing.
+///
+/// This class should not be used directly, and should be extended for different
+/// types of deltas.
+abstract class TextEditingDelta with TextEditingDeltaUtils {
+  /// Creates a delta for a given change to the editing state.
+  ///
+  /// {@template flutter.services.TextEditingDelta}
+  /// The [oldText], [deltaText], [deltaRange], [selection], and [composing]
+  /// arguments must not be null.
+  /// {@endtemplate}
+  const TextEditingDelta({
+    required this.oldText,
+    required this.deltaText,
+    required this.deltaRange,
+    required this.selection,
+    required this.composing,
+  }) : assert(oldText != null),
+       assert(deltaText != null),
+       assert(deltaRange != null),
+       assert(selection != null),
+       assert(composing != null);
+
+  /// Creates an instance of this class from a JSON object by checking the
+  /// deltaType sent by the engine and building the appropriate delta.
+  factory TextEditingDelta.fromJSON(Map<String, dynamic> encoded) {
+    switch (encoded['deltaType'] as String) {
+      case 'TextEditingDeltaType.insertion':
+        return TextEditingDeltaInsertion.fromJSON(encoded);
+      case 'TextEditingDeltaType.deletion':
+        return TextEditingDeltaDeletion.fromJSON(encoded);
+      case 'TextEditingDeltaType.replacement':
+        return TextEditingDeltaReplacement.fromJSON(encoded);
+      case 'TextEditingDeltaType.equality':
+        return TextEditingDeltaEquality.fromJSON(encoded);
+      default:
+        return TextEditingDeltaEquality.fromJSON(encoded);
```

Comment:
nit: maybe `assert(false)` here if the `default` clause should not be reachable.

Reply:
Good point